app/src/main/java/com/github/mateo762/myapplication/upload_gallery/UploadPictureActivity.kt

```diff
@@ -0,0 +1,160 @@
+import com.github.mateo762.myapplication.databinding.UploadPictureBinding
+import com.google.firebase.auth.ktx.auth
+import com.google.firebase.database.DataSnapshot
+import com.google.firebase.database.DatabaseError
+import com.google.firebase.database.ValueEventListener
+import com.google.firebase.database.ktx.database
+import com.google.firebase.ktx.Firebase
+import com.google.firebase.storage.FirebaseStorage
+import java.text.SimpleDateFormat
+import java.util.*
+
+class UploadPictureActivity : AppCompatActivity() {
+
+    private lateinit var binding: UploadPictureBinding
+    private lateinit var imageUri: Uri
+
+    private var defaultUserString: String = "testUser"
+    private var currentUser: String = defaultUserString
+
+    lateinit var imagesList: ArrayList<UserImage>
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        binding = UploadPictureBinding.inflate(layoutInflater)
+        setContentView(binding.root)
+
+        binding.selectPictureButton.setOnClickListener {
+            selectImage()
+        }
+
+        binding.uploadPictureButton.setOnClickListener {
+            uploadImage()
+        }
+
+        binding.imageRecycler.layoutManager = LinearLayoutManager(this)
+        imagesList = arrayListOf()
+
+        //Populate the recycler view with the downloaded images from the current user firebase database
+        var currentUser = Firebase.auth.currentUser?.displayName
+
+        if (currentUser == null) {
+            currentUser = defaultUserString
+        }
+
+        val db = Firebase.database.reference.child("users").child(currentUser)
+        db.addValueEventListener(object : ValueEventListener {
+            override fun onDataChange(snapshot: DataSnapshot) {
+                imagesList = arrayListOf()
+                if (snapshot.exists()) {
+                    for (dataSnapshot in snapshot.children) {
+                        val image = dataSnapshot.getValue(UserImage::class.java)
+                        imagesList.add(image!!)
+                    }
+
+                    binding.imageRecycler.adapter =
+                        ImageAdapter(imagesList, this@UploadPictureActivity)
+                }
+            }
+
+            override fun onCancelled(error: DatabaseError) {
+                Toast.makeText(this@UploadPictureActivity, error.toString(), Toast.LENGTH_SHORT)
+                    .show()
+            }
+        })
+
+
+    }
+
+    //Select an image from the gallery
+    private fun selectImage() {
+        val intent = Intent()
+        intent.type = "image/*"
+        intent.action = Intent.ACTION_GET_CONTENT
+
+        startActivityForResult(intent, 100)
+    }
+
+    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+        super.onActivityResult(requestCode, resultCode, data)
+
+        if (requestCode == 100 && resultCode == RESULT_OK) {
+            imageUri = data?.data!!
```

Comment:
!! will crash the app if it is null, we should handle the case when it is null with a Toast for example. A simple null check inside an if-else block will fix the problem